### PR TITLE
#244 Support function to read and write detachable links to/form a pi…

### PR DESCRIPTION
**Pull request by tomlyn**
…pelineFlow

Fixes: #244 
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Comment by caritaou** [comment]
There are a lot of console errors when loading the detachedLinksCanvas if I don't select the "Enable detach link: option in the test harness first. But I think this is fine
image

**Comment by caritaou** [comment]
Theres an error on the console if I drop a node on a detached link
image

File: canvas_modules/common-canvas/src/common-canvas/svg-canvas-renderer.js [hunk 1]
```diff
@@ -3716,19 +3706,31 @@ export default class SVGCanvasRenderer {
...
 	}
 
-	completeNewLink(trgNode) {
+	// Handles the completion of a new link being drawn from a source node.
+	completeNewLink() {
+		var trgNode = this.getNodeAtMousePos();
+		if (trgNode !== null) {
+			this.completeNewLinkOnNode(trgNode);
+		} else {
+			if (this.config.enableDetachableLinks &&
```
[the remark this answers is not in the document]
Comment (tomlyn) [reply canvas_modules/common-canvas/src/common-canvas/svg-canvas-renderer.js:3722]: This logic has changed in the latest code. The if in the else has an else of its own.

File: canvas_modules/common-canvas/src/common-canvas/svg-canvas-renderer.js [hunk 2]
```diff
@@ -3716,19 +3706,31 @@ export default class SVGCanvasRenderer {
...
+					this.drawingNewLinkData.action === "node-node" &&
+					!this.config.enableAssocLinkCreation) {
+				this.completeNewDetachedLink();
+			} else {
+				this.stopDrawingNewLink();
+			}
+		}
+	}
+
+	// Handles the completion of a new link when the end if dropped on a node.
```
[the remark this answers is not in the document]
Comment (tomlyn) [reply canvas_modules/common-canvas/src/common-canvas/svg-canvas-renderer.js:3732]: Yes, fixed, thanks.

File: canvas_modules/common-canvas/src/common-canvas/svg-canvas-renderer.js [hunk 3]
```diff
@@ -3764,6 +3766,31 @@ export default class SVGCanvasRenderer {
 		this.drawingNewLinkData = null;
 	}
 
+	// Handles the completion of a new link when the end if dropped away from
```
[the remark this answers is not in the document]
Comment (tomlyn) [reply canvas_modules/common-canvas/src/common-canvas/svg-canvas-renderer.js:3769]: Yes, fixed, thanks.

**Comment by tomlyn** [comment]
In answer to your comments:
1) The errors that are displayed, when `enableDetachableLinks` is not, set are correct in that any detachable links in the pipelineFlow cannot be drawn when then option is not switched on.  The error detection code has been there in one form another for a number of years to catch any invalid links in the pipelineFlow so I just left it as is. I think it should be OK since these errors will only appear if you try to display a pipelineFlow that is not compatible with your config options. 
2) I fixed the code to allow prevent a node being dropped on a detached link when `enableInsertNodeDroppedOnLink` is set. This was the preferred option of the Data Stage team.

**Pull request merged**